### ISSUE TITLE
Consider always resetting SemanticNode in markNeedsSemanticsUpdate

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#11775]
Background: In https://github.com/flutter/flutter/pull/11770 it was very difficult to concisely describe the API contract of `markNeedsSemanticsUpdate` and `SemanticsAnnotator`.

Currently, `markNeedsSemanticsUpdate` will reset the corresponding SemanticsNode, if `onlyChanges: false` is passed and it will not reset the SemanticsNode if `onlyChanges: true` is passed (in addition to that the flag also does other things, see below). Depending on the value of the flag, the SemanticsAnnotator might get either a freshly reseted node to annotate or a node that still has the previous annotation persisted. This creates a lot of mental overhead on two fronts:
1. When calling `markNeedsSemanticsUpdate` I need to decide if I require the SemanticsNode to be reseted.
2. When implementing my SemanticsAnnotator I need to make sure it works correctly in both cases (reseted node and node with persisted annotations).

To complicate matters even further, the flag `onlyChanges` is overloaded: It also needs to be set to `false` when the shape of the semantic tree changes in any way.

Looking at the relevant code bits, I think we can simplify this by always resetting the corresponding SemanticsNode and guaranteeing that semantics annotations are never persisted between calls to the SemanticsAnnotator. This simplifies two things:
1. The flag `onlyChanges` exclusively controls whether the semantics update causes changes to the shape of the semantics tree or not (we could rename to `localChangesOnly` to indicate that a RenderObject only intents to change values on the local SemanticsNode without affecting any other nodes in the tree).
2. The SemanticsAnnotator will always get a clean SemanticsNode.

What's the cost?
Always resetting the SemanticsNode will mean that the node will always end up in the dirty list [1]. However, after a RenderObject calls `markNeedsSemanticsUpdate` it should end up there anyway because with this call the RenderObject tells us that it wants to change something on its SemanticsNode and any change will put the node in the dirty list (for example [2]). There is only additional cost involved in always resetting the node if after calling `markNeedsSemanticsUpdate` the RenderObject does not change anything on the node, which I would consider a bug. `markNeedsSemanticsUpdate`should not have been called in the first place.

There is also the cost of the reset itself, which is just a handful of assignments and when we set a value on the reseted note, we'll have to do one extra assignment that would not have happened on an unreseted note (for example [2]). I think both of these extra costs are neglect-able.

To sum it up, for almost no cost we would get a much cleaner API: the flag `onlyChanges` gets one single meaning and stops being overloaden. And SemanticsAnnotators will only have to deal with rested SemanticsNodes.

/cc @Hixie for comment.

[1] https://github.com/flutter/flutter/blob/f969b777b54aa0f72320a4ddfde0883ed300283e/packages/flutter/lib/src/rendering/semantics.dart#L385:L393
[2] https://github.com/flutter/flutter/blob/f969b777b54aa0f72320a4ddfde0883ed300283e/packages/flutter/lib/src/rendering/semantics.dart#L345:L351

**Comment by Hixie** [comment]
I believe the idea behind the original design was that without the reset, only one RenderObject would update the SemanticsNode, and with the reset, we'd have to walk a huge chunk of the tree.

Anyway, I'm fine with any and all simplifications of this code, so long as we verify that it doesn't negatively impact performance.

**Comment by goderbauer** [comment]
> I believe the idea behind the original design was that without the reset, only one RenderObject would update the SemanticsNode, and with the reset, we'd have to walk a huge chunk of the tree.

In the current impl, we always walk at least the render tree between two semantic nodes. That doesn't change.

Performance numbers for the benchmarks we have on my local system are in the same range.

**Comment by Hixie** [comment]
Excellent. Make it so!

**Comment by goderbauer** [comment]
It's waiting for review: https://github.com/flutter/flutter/pull/11770 :)
